### PR TITLE
vLLM request parameters `constraint_mode` and `constraint`

### PR DESCRIPTION
**Pull request by viktor-ferenczi**
Introduced generic vLLM request parameters `constraint_mode` and `constraint` to replace `json_schema` and `regex`.

The main reason for the change is to allow for adding an arbitrary number of grammars types without having to add new parameters to the vLLM API every time. It also reduces the probability of a future collision with an original vLLM request parameter name. 

It also reduces confusion due to naming, like `cfg` meaning both "configuration" and "Context-Free Grammar", which was confusing at least to me. (This is for the upcoming PR to introduce full Lark grammar support to the vLLM API.)

Kept the old `json_schema` and `regex` parameters for backwards compatibility, but we should remove them soon.

### CODE REVIEW

**Comment by viktor-ferenczi** [comment]
We could also name these `constraint_mode` and `constraint`, then use `"grammar"` as a `constraint_mode` value for the full Lark grammar. What do you think about these names?

**Comment by viktor-ferenczi** [comment]
@rlouf Could you please review this? Thanks!

**Comment by rlouf** [comment]
I find this naming less explicit than the current one. It may be confusing.

**Comment by viktor-ferenczi** [comment]
Fair enough. We'll see whether this will still be the case after adding the 5th logits processor.

**Comment by rlouf** [comment]
We'll cross that bridge when we come to it 😁
